File: Examples/Python/Containers/MongoDB-REST/src/services/MongoDB/REST/mongodb_rest_client.py
import json
from bson import ObjectId
import os
import sys
from typing import Optional, Any

from src.services.MongoDB.CRUD.mongodb_crud_client import MongoDB_CRUD_Client
from src.services.MongoDB.CRUD.mongodb_config import MongoDB_Config
from .rest_response import REST_Response

class MongoDB_REST_Client:
    def __init__(self, mongodb_config: MongoDB_Config):
        self.__mongodb_config = mongodb_config
        self.__crud_client = None

    def __bad_request(self, message: str) -> REST_Response:
        return REST_Response(400, 'text/plain', message)

    def __error_message(self, e: Exception) -> REST_Response:
        return self.__error_message(str(e))
    
    def __error_message(self, message: str) -> REST_Response:
        return REST_Response(500, 'text/plain', message)

    def __not_found(self, id: ObjectId) -> REST_Response:
        return REST_Response(404, 'text/plain', f'Error: Document with id {id} not found')

    def __ok(self, data: dict) -> REST_Response:
        return REST_Response(200, 'application/json', json.dumps(data, default=str))
    
    def __created(self, id: ObjectId) -> REST_Response:
        return REST_Response(201, 'text/plain', str(id))
    
    def __deleted(self, id: str) -> REST_Response:
        return REST_Response(200, 'text/plain', f'Document with id {id} deleted')
    
    def __verify_parameters(self, db_name: Optional[str], collection_name: Optional[str], id: Optional[str], is_id_required: bool, data: Optional[Any], is_data_required: bool) -> Optional[dict]:
        if db_name is None:
            return self.__bad_request('Error: Missing db_name parameter')
        if collection_name is None:
            return self.__bad_request('Error: Missing collection_name parameter')
        if is_id_required:
            if id is None:
                return self.__bad_request('Error: Missing id parameter')
            elif not ObjectId.is_valid(id):
                return self.__bad_request('Error: Invalid id')
        if is_data_required and data is None:
            return self.__bad_request('Error: Missing body')
        return None
    
    def __get_client(self) -> MongoDB_CRUD_Client:
        if self.__crud_client is None:
            self.__crud_client = MongoDB_CRUD_Client(self.__mongodb_config)
        return self.__crud_client
    
    def ping(self):
        with self.__get_client() as client:
            client.ping()

    def get(self, db_name: Optional[str], collection_name: Optional[str], id: Optional[str]) -> REST_Response: 
        try:
            error = self.__verify_parameters(db_name, collection_name, id, True, None, False)
            if error is not None:
                return error
            
            with self.__get_client() as client:
                document = client.read(db_name, collection_name, ObjectId(id))
            if document is not None:
                return self.__ok(document)
            return self.__not_found(id)
        except Exception as e:
            return self.__error_message(e)
        
    def post(self, db_name: Optional[str], collection_name: Optional[str], data: Optional[any]) -> REST_Response:
        try:
            error = self.__verify_parameters(db_name, collection_name, None, False, data, True)
            if error is not None:
                return error
            
            with self.__get_client() as client:
                document_id = client.create(db_name, collection_name, data)
            if document_id is not None:
                return self.__created(document_id)
            return self.__error_message('Failed to create document')
        except Exception as e:
            return self.__error_message(e)

    def put(self, db_name: Optional[str], collection_name: Optional[str], id: Optional[str], data: Optional[any]) -> REST_Response: 
        try:
            error = self.__verify_parameters(db_name, collection_name, id, True, data, True)
            if error is not None:
                return error
            
            with self.__get_client() as client:
                document = client.update(db_name, collection_name, ObjectId(id), data)
            if document is not None:
                return self.__ok(document)
            return self.__not_found(id)
        except Exception as e:
            return self.__error_message(e)

    def delete(self, db_name: Optional[str], collection_name: Optional[str], id: Optional[str]) -> REST_Response: 
        try:
            error = self.__verify_parameters(db_name, collection_name, id, True, None, False)
            if error is not None:
                return error
            
            with self.__get_client() as client:
                result = client.delete(db_name, collection_name, ObjectId(id))
            if result:
                return self.__deleted(id)
            return self.__not_found(id)
        except Exception as e:
            return self.__error_message(e)

File: Examples/Rust/Containers/MongoDB-REST/src/src/services/mongodb/rest/mongodb_rest_client.rs
use mongodb::{bson::{ oid::ObjectId}};
use serde_json::json;

use crate::services::mongodb::crud::mongodb_crud_client::MongoDBCRUDClient;
use crate::services::mongodb::rest::rest_response::RESTResponse;

pub struct MongoDBRESTClient {
    crud_client: Option<MongoDBCRUDClient>,
}

impl MongoDBRESTClient {
    pub fn new(crud_client: MongoDBCRUDClient) -> Self {
        Self {
            crud_client: Some(crud_client),
        }
    }

    fn get_client(&self) -> &MongoDBCRUDClient {
        self.crud_client.as_ref().expect("MongoDB CRUD client is not set")
    }

    fn bad_request(&self, message: &str) -> RESTResponse {
        RESTResponse::new(400, "text/plain", message.to_string())
    }

    fn error_message(&self, message: &str) -> RESTResponse {
        RESTResponse::new(500, "text/plain", message.to_string())
    }

    fn not_found(&self, id: &ObjectId) -> RESTResponse {
        RESTResponse::new(404, "text/plain", format!("Error: Document with id {} not found", id))
    }

    fn ok(&self, data: serde_json::Value) -> RESTResponse {
        RESTResponse::new(200, "application/json", data.to_string())
    }

    fn created(&self, id: &ObjectId) -> RESTResponse {
        RESTResponse::new(201, "text/plain", id.to_string())
    }

    fn deleted(&self, id: &str) -> RESTResponse {
        RESTResponse::new(200, "text/plain", format!("Document with id {} deleted", id))
    }

    fn verify_parameters(
        &self,
        db_name: Option<&str>,
        collection_name: Option<&str>,
        id: Option<&str>,
        is_id_required: bool,
        data: Option<&serde_json::Value>,
        is_data_required: bool,
    ) -> Option<RESTResponse> {
        if db_name.is_none() {
            return Some(self.bad_request("Error: Missing db_name parameter"));
        }
        if collection_name.is_none() {
            return Some(self.bad_request("Error: Missing collection_name parameter"));
        }
        if is_id_required {
            if id.is_none() {
                return Some(self.bad_request("Error: Missing id parameter"));
            } else if !ObjectId::parse_str(id.unwrap()).is_ok() {
                return Some(self.bad_request("Error: Invalid id"));
            }
        }
        if is_data_required && data.is_none() {
            return Some(self.bad_request("Error: Missing body"));
        }
        None
    }

    pub async fn ping(&self) -> Result<(), mongodb::error::Error> {
        match self.get_client().ping().await {
            Ok(..) => Ok(()),
            Err(e) => Err(e),
        }
    }

    pub async fn get(
        &self,
        db_name: Option<&str>,
        collection_name: Option<&str>,
        id: Option<&str>,
    ) -> RESTResponse {
        if let Some(error) = self.verify_parameters(db_name, collection_name, id, true, None, false) {
            return error;
        }

        let client = self.get_client();
        match client.read(db_name.unwrap(), collection_name.unwrap(), ObjectId::parse_str(id.unwrap()).unwrap()).await {
            Ok(Some(document)) => self.ok(json!(document)),
            Ok(None) => self.not_found(&ObjectId::parse_str(id.unwrap()).unwrap()),
            Err(e) => self.error_message(&e.to_string()),
        }
    }

    pub async fn post(
        &self,
        db_name: Option<&str>,
        collection_name: Option<&str>,
        data: Option<serde_json::Value>,
    ) -> RESTResponse {
        if let Some(error) = self.verify_parameters(db_name, collection_name, None, false, data.as_ref(), true) {
            return error;
        }

        let client = self.get_client();
        match client.create(db_name.unwrap(), collection_name.unwrap(), data.unwrap()).await {
            Ok(document_id) => self.created(&document_id),
            //Ok(None) => self.error_message("Failed to create document"),
            Err(e) => self.error_message(&e.to_string()),
        }
    }

    pub async fn put(
        &self,
        db_name: Option<&str>,
        collection_name: Option<&str>,
        id: Option<&str>,
        data: Option<serde_json::Value>,
    ) -> RESTResponse {
        if let Some(error) = self.verify_parameters(db_name, collection_name, id, true, data.as_ref(), true) {
            return error;
        }

        let client = self.get_client();
        match client.update(db_name.unwrap(), collection_name.unwrap(), ObjectId::parse_str(id.unwrap()).unwrap(), data.unwrap()).await {
            Ok(Some(document)) => self.ok(json!(document)),
            Ok(None) => self.not_found(&ObjectId::parse_str(id.unwrap()).unwrap()),
            Err(e) => self.error_message(&e.to_string()),
        }
    }

    pub async fn delete(
        &self,
        db_name: Option<&str>,
        collection_name: Option<&str>,
        id: Option<&str>,
    ) -> RESTResponse {
        if let Some(error) = self.verify_parameters(db_name, collection_name, id, true, None, false) {
            return error;
        }

        let client = self.get_client();
        match client.delete(db_name.unwrap(), collection_name.unwrap(), ObjectId::parse_str(id.unwrap()).unwrap()).await {
            Ok(true) => self.deleted(id.unwrap()),
            Ok(false) => self.not_found(&ObjectId::parse_str(id.unwrap()).unwrap()),
            Err(e) => self.error_message(&e.to_string()),
        }
    }
}
